### PR TITLE
kernel: do not call swap if next ready thread is current thread.

### PR DESCRIPTION
**Pull request by nashif**
Check if next ready thread is same as current thread before calling z_swap.
This avoids calling swap to just go back to the original thread.

Original code: thread 0x20000118 switches out and then in again...

>> 0x20000118 gives semaphore(signal): 0x20000104 (count: 0)
>> thread ready: 0x20000118
>> 0x20000118 switched out
>> 0x20000118 switched in
>> end call to k_sem_give
>> 0x20000118 takes semaphore(wait): 0x200000f4 (count: 0)
>> thread pend: 0x20000118
>> 0x20000118 switched out
>> 0x200001d0 switched in

with this patch:

>> 0x200001d0 gives semaphore(signal): 0x200000f4 (count: 0)
>> thread ready: 0x200001d0
>> end call to k_sem_give
>> 0x200001d0 takes semaphore(wait): 0x20000104 (count: 0)
>> thread pend: 0x200001d0
>> 0x200001d0 switched out
>> 0x20000118 switched in
>> end call to k_sem_take

The above is output from tracing with a custom format used for debugging.

### CODE REVIEW

File: kernel/sched.c [hunk 1]
```diff
@@ -748,9 +748,27 @@ static inline int resched(uint32_t key)
...
 
+/*
+ * Check if the next ready thread is the same as the current thread
+ * and save the trip if true.
+ */
+static inline bool need_swap(void)
+{
+	/* the SMP case will be handled in C based z_swap() */
+#ifdef CONFIG_SMP
+	return 1;
```
Comment (ceolin) [inline kernel/sched.c:759]: Should return a boolean and not an int here for the misra c sake

**Review by ceolin — APPROVED** [review]
just a minor thing

**Review by dcpleung — APPROVED** [review]

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by andrewboie** [comment]
should get @andyross's blessing but this LGTM

**Review by andyross — APPROVED** [review]
Sorry,  I blessed this in a chat window and forgot to +1 the PR

**Pull request merged**